File: lib/quota.dart
/// Page allowance for one user, as reported by the Worker.
///
/// Free: 100 pages per calendar month (reset, no rollover).
/// Subscribed: unlimited* — 5000 guaranteed/month with 500/day and
/// 10000/month guardrails.
class QuotaStatus {
  const QuotaStatus({
    required this.freeUsed,
    required this.freeTotal,
    required this.paidBalance,
    required this.pro,
    required this.unlimited,
    required this.paidUsed,
    required this.paidCap,
    required this.dailyUsed,
    required this.dailyCap,
    required this.month,
  });

  factory QuotaStatus.fromJson(Map<String, dynamic> json) {
    int asInt(Object? v) => v is int ? v : int.tryParse('$v') ?? 0;
    bool asBool(Object? v) => v == true;
    return QuotaStatus(
      freeUsed: asInt(json['freeUsed']),
      freeTotal: asInt(json['freeTotal']),
      paidBalance: asInt(json['paidBalance']),
      pro: asBool(json['pro']),
      unlimited: asBool(json['unlimited']),
      paidUsed: asInt(json['paidUsed']),
      paidCap: asInt(json['paidCap']),
      dailyUsed: asInt(json['dailyUsed']),
      dailyCap: asInt(json['dailyCap']),
      month: '${json['month'] ?? ''}',
    );
  }

  /// Monthly free pages consumed.
  final int freeUsed;

  /// Free pages per month (100).
  final int freeTotal;

  /// Legacy compat: remaining monthly allowance.
  final int paidBalance;

  /// RevenueCat `pro_pages` entitlement currently active (unused;
  /// the server is the source of truth for access).
  final bool pro;

  /// True while a subscription is live.
  final bool unlimited;

  /// Paid pages consumed this calendar month.
  final int paidUsed;

  /// Paid hard cap per month (10000).
  final int paidCap;

  /// Paid pages consumed today (UTC).
  final int dailyUsed;

  /// Paid daily cap (500).
  final int dailyCap;

  /// Current billing window, YYYY-MM.
  final String month;

  int get freeLeft => freeTotal - freeUsed < 0 ? 0 : freeTotal - freeUsed;

  bool get canSummarize {
    if (unlimited) {
      return dailyUsed < dailyCap && paidUsed < paidCap;
    }
    return freeLeft > 0;
  }

  /// Why summarizing is blocked right now (null when allowed):
  /// free exhaustion → paywall; daily/monthly caps → wait it out.
  String? get denial {
    if (canSummarize) return null;
    if (!unlimited) return 'free';
    if (dailyUsed >= dailyCap) return 'daily';
    return 'monthly';
  }

  String get denialMessage {
    if (denial == null) return '';
    return switch (denial) {
      'daily' => 'Daily limit reached — new pages tomorrow.',
      'monthly' => 'Monthly cap reached — new pages next month.',
      _ => 'Out of pages — subscribe for unlimited.',
    };
  }

  /// Pill label: used/total — "5/100" free, "5/∞" unlimited.
  String get pillLabel =>
      unlimited ? '$paidUsed/∞' : '$freeUsed/$freeTotal';
}

File: pwa/src/quota.ts
/**
 * Page allowance for one user, as reported by the Worker.
 * Field-for-field port of the old QuotaStatus — the server is source
 * of truth, these strings are the UX contract. Copy, don't reinterpret.
 *
 * Free: 100 pages per calendar month (reset, no rollover).
 * Subscribed: unlimited* — 5000 guaranteed/month with 500/day and
 * 10000/month guardrails.
 */

export type Denial = 'free' | 'daily' | 'monthly';

export interface QuotaJson {
  freeUsed: unknown;
  freeTotal: unknown;
  paidBalance: unknown;
  pro: unknown;
  unlimited: unknown;
  paidUsed: unknown;
  paidCap: unknown;
  dailyUsed: unknown;
  dailyCap: unknown;
  month: unknown;
}

function asInt(v: unknown): number {
  if (typeof v === 'number' && Number.isFinite(v)) return Math.trunc(v);
  const n = parseInt(`${v ?? ''}`, 10);
  return Number.isFinite(n) ? n : 0;
}

export class QuotaStatus {
  constructor(
    readonly freeUsed: number,
    readonly freeTotal: number,
    readonly paidBalance: number,
    readonly pro: boolean,
    readonly unlimited: boolean,
    readonly paidUsed: number,
    readonly paidCap: number,
    readonly dailyUsed: number,
    readonly dailyCap: number,
    readonly month: string,
  ) {}

  static fromJson(json: QuotaJson): QuotaStatus {
    return new QuotaStatus(
      asInt(json.freeUsed),
      asInt(json.freeTotal),
      asInt(json.paidBalance),
      json.pro === true,
      json.unlimited === true,
      asInt(json.paidUsed),
      asInt(json.paidCap),
      asInt(json.dailyUsed),
      asInt(json.dailyCap),
      `${json.month ?? ''}`,
    );
  }

  get freeLeft(): number {
    const left = this.freeTotal - this.freeUsed;
    return left < 0 ? 0 : left;
  }

  get canSummarize(): boolean {
    if (this.unlimited) {
      return this.dailyUsed < this.dailyCap && this.paidUsed < this.paidCap;
    }
    return this.freeLeft > 0;
  }

  /** Why summarizing is blocked right now (null when allowed). */
  get denial(): Denial | null {
    if (this.canSummarize) return null;
    if (!this.unlimited) return 'free';
    if (this.dailyUsed >= this.dailyCap) return 'daily';
    return 'monthly';
  }

  get denialMessage(): string {
    if (this.denial === null) return '';
    switch (this.denial) {
      case 'daily':
        return 'Daily limit reached — new pages tomorrow.';
      case 'monthly':
        return 'Monthly cap reached — new pages next month.';
      default:
        return 'Out of pages — subscribe for unlimited.';
    }
  }

  /** Pill label: used/total — "5/100" free, "5/∞" unlimited. */
  get pillLabel(): string {
    return this.unlimited ? `${this.paidUsed}/∞` : `${this.freeUsed}/${this.freeTotal}`;
  }
}
